File: app/src/main/java/br/com/pedrosilva/tecnonutri/presentation/ui/adapters/FoodAdapter.java
package br.com.pedrosilva.tecnonutri.presentation.ui.adapters;

import android.content.Context;
import android.content.res.Resources;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import br.com.pedrosilva.tecnonutri.R;
import br.com.pedrosilva.tecnonutri.domain.entities.FeedItem;
import br.com.pedrosilva.tecnonutri.domain.entities.Food;
import br.com.pedrosilva.tecnonutri.util.AppUtil;

/**
 * Created by psilva on 3/21/17.
 */

public class FoodAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final Context context;

    private final int VIEW_TYPE_FOOD = 0;
    private final int VIEW_TYPE_TOTAL = 1;

    private FeedItem feedItem;

    public FoodAdapter(Context context) {
        this.context = context;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());

        RecyclerView.ViewHolder holder;
        View view;

        if (viewType == VIEW_TYPE_FOOD) {
            view = inflater.inflate(R.layout.item_food, parent, false);
            holder = new ViewHolderFood(view);
        } else {
            view = inflater.inflate(R.layout.item_food_total, parent, false);
            holder = new ViewHolderTotal(view);
        }
        return holder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        switch (holder.getItemViewType()) {
            case VIEW_TYPE_FOOD:
                ViewHolderFood holderFood = (ViewHolderFood) holder;
                holderFood.setup(getItem(position));
                break;
            case VIEW_TYPE_TOTAL:
                ViewHolderTotal holderTotal = (ViewHolderTotal) holder;
                holderTotal.setup(feedItem);
                break;
        }
    }

    private Food getItem(int position) {
        if (feedItem != null) {
            return feedItem.getFoods().get(position);
        }
        return null;
    }

    @Override
    public int getItemViewType(int position) {
        if (position < getItemCount() - 1) {
            return VIEW_TYPE_FOOD;
        }
        return VIEW_TYPE_TOTAL;
    }

    public void setFoodItem(FeedItem feedItem) {
        this.feedItem = feedItem;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        if (feedItem != null) {
            return feedItem.getFoods().size() + 1;
        }
        return 0;
    }

    private class ViewHolderFood extends RecyclerView.ViewHolder {

        private TextView tvDescription;
        private TextView tvQty;

        private TextView tvNutritionalEnergy;
        private TextView tvNutritionalCarbohydrate;
        private TextView tvNutritionalProtein;
        private TextView tvNutritionalFat;

        public ViewHolderFood(View v) {
            super(v);
            bindElements(v);
        }

        private void bindElements(View v) {
            tvDescription = (TextView) v.findViewById(R.id.tv_description);
            tvQty = (TextView) v.findViewById(R.id.tv_qty);

            tvNutritionalEnergy = (TextView) v.findViewById(R.id.tv_nutritional_energy);
            tvNutritionalCarbohydrate = (TextView) v.findViewById(R.id.tv_nutritional_carbohydrate);
            tvNutritionalProtein = (TextView) v.findViewById(R.id.tv_nutritional_protein);
            tvNutritionalFat = (TextView) v.findViewById(R.id.tv_nutritional_fat);
        }

        private void setup(Food food) {
            final Context context = itemView.getContext();
            final Resources resources = context.getResources();

            tvDescription.setText(food.getDescription());

            String qty = resources.getString(R.string.qty_food, AppUtil.formatDecimal(food.getAmount()), food.getMeasure(), AppUtil.formatWeight(food.getWeight()));
            tvQty.setText(qty);

            String kcal = resources.getString(R.string.qty_energy, AppUtil.formatKcal(food.getEnergy()));
            tvNutritionalEnergy.setText(kcal);

            String carb = resources.getString(R.string.weight, AppUtil.formatWeight(food.getCarbohydrate()));
            tvNutritionalCarbohydrate.setText(carb);

            String prot = resources.getString(R.string.weight, AppUtil.formatWeight(food.getProtein()));
            tvNutritionalProtein.setText(prot);

            String fat = resources.getString(R.string.weight, AppUtil.formatWeight(food.getFat()));
            tvNutritionalFat.setText(fat);
        }
    }

    private class ViewHolderTotal extends RecyclerView.ViewHolder {

        private TextView tvNutritionalEnergy;
        private TextView tvNutritionalCarbohydrate;
        private TextView tvNutritionalProtein;
        private TextView tvNutritionalFat;

        public ViewHolderTotal(View v) {
            super(v);
            bindElements(v);
        }

        private void bindElements(View v) {
            tvNutritionalEnergy = (TextView) v.findViewById(R.id.tv_nutritional_energy);
            tvNutritionalCarbohydrate = (TextView) v.findViewById(R.id.tv_nutritional_carbohydrate);
            tvNutritionalProtein = (TextView) v.findViewById(R.id.tv_nutritional_protein);
            tvNutritionalFat = (TextView) v.findViewById(R.id.tv_nutritional_fat);
        }

        private void setup(FeedItem feedItem) {
            final Context context = itemView.getContext();
            final Resources resources = context.getResources();

            String kcal = resources.getString(R.string.qty_energy, AppUtil.formatKcal(feedItem.getEnergy()));
            tvNutritionalEnergy.setText(kcal);

            String carb = resources.getString(R.string.weight, AppUtil.formatWeight(feedItem.getCarbohydrate()));
            tvNutritionalCarbohydrate.setText(carb);

            String prot = resources.getString(R.string.weight, AppUtil.formatWeight(feedItem.getProtein()));
            tvNutritionalProtein.setText(prot);

            String fat = resources.getString(R.string.weight, AppUtil.formatWeight(feedItem.getFat()));
            tvNutritionalFat.setText(fat);
        }
    }
}

File: app/src/main/kotlin/br/com/pedrosilva/tecnonutri/presentation/ui/adapters/FoodAdapter.kt
package br.com.pedrosilva.tecnonutri.presentation.ui.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import br.com.pedrosilva.tecnonutri.R
import com.pedrenrique.tecnonutri.domain.FeedItem
import com.pedrenrique.tecnonutri.domain.Food
import br.com.pedrosilva.tecnonutri.util.AppUtil
import kotlinx.android.synthetic.main.item_food.view.tv_description
import kotlinx.android.synthetic.main.item_food.view.tv_qty
import kotlinx.android.synthetic.main.item_food_nutritional.view.tv_nutritional_carbohydrate
import kotlinx.android.synthetic.main.item_food_nutritional.view.tv_nutritional_energy
import kotlinx.android.synthetic.main.item_food_nutritional.view.tv_nutritional_fat
import kotlinx.android.synthetic.main.item_food_nutritional.view.tv_nutritional_protein

class FoodAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    var feedItem: FeedItem? = null
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    private val items: List<Food>
        get() = feedItem?.foods ?: listOf()

    companion object {
        private const val VIEW_TYPE_FOOD = 0
        private const val VIEW_TYPE_TOTAL = 1
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val holder: RecyclerView.ViewHolder
        val view: View
        if (viewType == VIEW_TYPE_FOOD) {
            view = inflater.inflate(R.layout.item_food, parent, false)
            holder = ViewHolderFood(view)
        } else {
            view = inflater.inflate(R.layout.item_food_total, parent, false)
            holder = ViewHolderTotal(view)
        }
        return holder
    }

    override fun onBindViewHolder(
        holder: RecyclerView.ViewHolder,
        position: Int
    ) {
        when (holder.itemViewType) {
            VIEW_TYPE_FOOD -> {
                val holderFood = holder as ViewHolderFood
                holderFood.setup(getItem(position))
            }
            VIEW_TYPE_TOTAL -> feedItem?.let { item ->
                val holderTotal = holder as ViewHolderTotal
                holderTotal.setup(item)
            }
        }
    }

    private fun getItem(position: Int): Food = items[position]

    override fun getItemViewType(position: Int): Int =
        if (position < itemCount - 1) VIEW_TYPE_FOOD
        else VIEW_TYPE_TOTAL

    override fun getItemCount(): Int =
        if (items.isNotEmpty()) items.size + 1
        else 0

    private inner class ViewHolderFood(view: View) : RecyclerView.ViewHolder(view) {

        fun setup(food: Food) = itemView.run {
            val resources = itemView.context.resources
            tv_description.text = food.description
            tv_qty.text = resources.getString(
                R.string.qty_food,
                AppUtil.formatDecimal(food.amount!!),
                food.measure,
                AppUtil.formatWeight(food.weight!!)
            )
            tv_nutritional_energy.text =
                resources.getString(R.string.qty_energy, AppUtil.formatKcal(food.energy!!))
            tv_nutritional_carbohydrate.text =
                resources.getString(R.string.weight, AppUtil.formatWeight(food.carbohydrate!!))
            tv_nutritional_protein.text =
                resources.getString(R.string.weight, AppUtil.formatWeight(food.protein!!))
            tv_nutritional_fat.text =
                resources.getString(R.string.weight, AppUtil.formatWeight(food.fat!!))
        }
    }

    private inner class ViewHolderTotal(view: View) : RecyclerView.ViewHolder(view) {

        fun setup(feedItem: FeedItem) = itemView.run {
            val resources = itemView.context.resources
            tv_nutritional_energy.text =
                resources.getString(R.string.qty_energy, AppUtil.formatKcal(feedItem.energy!!))
            tv_nutritional_carbohydrate.text =
                resources.getString(R.string.weight, AppUtil.formatWeight(feedItem.carbohydrate!!))
            tv_nutritional_protein.text =
                resources.getString(R.string.weight, AppUtil.formatWeight(feedItem.protein!!))
            tv_nutritional_fat.text =
                resources.getString(R.string.weight, AppUtil.formatWeight(feedItem.fat!!))
        }
    }
}